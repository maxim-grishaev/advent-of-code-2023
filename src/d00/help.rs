use crate::util;

pub fn read_thing() -> String {
    util::fs::read_file(file!(), ".txt")
}

pub fn parse_thing(str: String) -> u32 {
    0
}

#[cfg(test)]
pub mod t {
    // use insta::assert_debug_snapshot;

    use super::*;

    pub const STR: &str = r#"

    "#;

    fn get_test() -> String {
        STR.trim().to_string()
    }

    #[test]
    fn test_run() {
        assert_eq!(parse_thing(get_thing()), None);
    }
}
