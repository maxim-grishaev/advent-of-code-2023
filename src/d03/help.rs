use crate::util;

#[derive(Debug)]
pub struct NumPos {
    line_no: usize,
    col_no: usize,
    pub num: u32,
    digits_count: usize,
    // raw: Vec<char>,
}

fn val(line_no: usize, col_no: usize, num: Vec<char>) -> NumPos {
    NumPos {
        line_no: line_no.clone(),
        col_no: col_no.clone(),
        num: num.iter().collect::<String>().parse::<u32>().unwrap(),
        digits_count: num.len(),
        // raw: num,
    }
}

pub fn get_char_at(schema: &Vec<&str>, line_no: &isize, col_no: &isize) -> Option<char> {
    if line_no < &0 || col_no < &0 || *line_no >= schema.len() as isize {
        return None;
    }
    return schema[*line_no as usize].chars().nth(*col_no as usize);
}

pub fn is_adjacent(num: &NumPos, gear_coords: &(usize, usize)) -> bool {
    let distance = (num.line_no as isize - gear_coords.0 as isize).abs();
    if distance > 1 {
        return false;
    }
    let adjacent_coords = get_adjacent_coords(num);
    adjacent_coords.iter().any(|(line_no, col_no)| {
        *line_no == gear_coords.0 as isize && *col_no == gear_coords.1 as isize
    })
}

pub fn get_adjacent_coords(np: &NumPos) -> Vec<(isize, isize)> {
    let col_left: isize = np.col_no as isize - np.digits_count as isize - 1;
    let col_right = np.col_no as isize;

    let line_top = np.line_no as isize - 1;
    let line_cur = (np.line_no) as isize;
    let line_btm = (np.line_no + 1) as isize;

    let mut adjacent_coords: Vec<(isize, isize)> = vec![];

    // left vertical
    adjacent_coords.push((line_top, col_left));
    adjacent_coords.push((line_cur, col_left));
    adjacent_coords.push((line_btm, col_left));

    // horisontal
    (0..np.digits_count).for_each(|i| {
        adjacent_coords.push((line_top, col_left + 1 + i as isize));
        adjacent_coords.push((line_btm, col_left + 1 + i as isize));
    });

    // right vertical
    adjacent_coords.push((line_top, col_right));
    adjacent_coords.push((line_cur, col_right));
    adjacent_coords.push((line_btm, col_right));

    adjacent_coords
}

fn is_symbol(c: char) -> bool {
    !c.is_numeric() && c != '.'
}

pub fn is_detail(schema: &Vec<&str>, np: &NumPos) -> bool {
    get_adjacent_coords(np).iter().any(|(line_no, col_no)| {
        match get_char_at(schema, line_no, col_no) {
            Some(c) => is_symbol(c),
            None => return false,
        }
    })
}

pub fn find_all_numbers(schema: &Vec<&str>) -> Vec<NumPos> {
    let mut nums: Vec<NumPos> = vec![];
    loop {
        let (line_no, col_no) = match nums.last() {
            Some(last) => (last.line_no, last.col_no),
            None => (0, 0),
        };
        match find_number(schema, line_no, col_no) {
            Some(found) => {
                nums.push(found);
                continue;
            }
            None => return nums,
        }
    }
}

pub fn find_number(
    schema: &Vec<&str>,
    line_no_start: usize,
    col_no_start: usize,
) -> Option<NumPos> {
    if line_no_start >= schema.len() {
        return None;
    }
    let mut line_no = line_no_start;
    let mut col_no = col_no_start;
    let mut num: Vec<char> = vec![];

    // println!("start_pos: {:?} {:?}", line_no, col_no);

    let mut cur_line = schema[line_no].chars();
    // println!("line: {:?}", cur_line);
    if col_no > 0 {
        cur_line.nth(col_no - 1);
    }

    // println!("schema: {:?}", schema);
    // println!("cur_line: {:?}", cur_line);
    loop {
        // println!("loop: {:?} {:?} {:?}", line_no, col_no, cur_line);
        match cur_line.next() {
            Some(c) => {
                if c.is_numeric() {
                    num.push(c);
                    col_no += 1;
                    continue;
                }
                // end of number
                if num.len() > 0 {
                    // println!("end of number. num: {:?} @ {:?}", num, line_no);
                    return Some(val(line_no, col_no, num));
                }
                col_no += 1;
            }
            // end of line
            None => {
                if num.len() > 0 {
                    // println!("end of line. num: {:?} @ {:?}", num, line_no);
                    return Some(val(line_no, col_no, num));
                }

                // goto next line
                line_no += 1;
                col_no = 0;

                // end of schema?
                if line_no >= schema.len() {
                    return None;
                }
                // println!("NEXT LINE!");
                cur_line = schema[line_no].chars();
                // println!("line: {:?}", cur_line);
            }
        }
    }
    // None
}

pub fn read_schema() -> String {
    util::fs::read_file(file!(), "engine.txt")
}

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;

    const TEST_INPUT: &str = r#"
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
    "#;

    #[test]
    fn find_all() {
        let schema: Vec<&str> = TEST_INPUT.trim().lines().collect();
        let nums = find_all_numbers(&schema);
        assert_debug_snapshot!(nums);
    }

    #[test]
    fn get_adjacent_coords_snapshot() {
        let schema: Vec<&str> = TEST_INPUT.trim().lines().collect();
        let npo = find_number(&schema, 0, 0);
        let np = npo.unwrap();
        println!("{:?}", np);
        assert_debug_snapshot!(get_adjacent_coords(&np));
    }
}
