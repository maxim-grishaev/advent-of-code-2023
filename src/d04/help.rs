use crate::util;

#[derive(Debug)]
pub struct Scratchcard {
    pub name: String,
    pub win: Vec<u32>,
    pub have: Vec<u32>,
}

fn normalize_numbers(line: &str) -> Vec<u32> {
    line.split_whitespace()
        .map(|s| s.parse::<u32>().unwrap())
        .collect()
}

pub fn read_cards() -> Vec<Scratchcard> {
    let cards_str = util::fs::read_file(file!(), "cards.txt");
    cards_str
        .lines()
        .map(|s| {
            let mut line_split = s.split(':');
            let name = line_split.next().unwrap().to_string();
            let mut nums_split = line_split.next().unwrap().split("|");
            Scratchcard {
                name,
                win: normalize_numbers(nums_split.next().unwrap()),
                have: normalize_numbers(nums_split.next().unwrap()),
            }
        })
        .collect()
}

pub fn intersection(narr1: &Vec<u32>, narr2: &Vec<u32>) -> Vec<u32> {
    narr1
        .into_iter()
        .filter(|win_num| narr2.contains(win_num))
        .copied()
        .collect()
}
