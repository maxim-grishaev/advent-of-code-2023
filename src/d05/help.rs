use crate::util;

pub type ID = u64;
pub type AlmMapCfgLine = (ID, ID, ID);

#[derive(Debug)]
pub struct AlmMap {
    pub title: String,
    pub config: Vec<AlmMapCfgLine>,
}

#[derive(Debug)]
pub struct Almanac {
    pub seeds: Vec<ID>,
    pub maps: Vec<AlmMap>,
}

pub fn read_almanac() -> String {
    util::fs::read_file(file!(), "almanac.txt")
}

pub fn parse_almanac(alm_str: &str) -> Almanac {
    let mut sections = alm_str.split("\n\n");

    let mut seeds_split = sections
        .next()
        .unwrap() // seeds line: "seeds: 79 14 55 13"
        .split(":");
    seeds_split.next(); // seeds label: "seeds"
    let seeds = seeds_split
        .next()
        .unwrap() // seeds string: "79 14 55 13"
        .split_whitespace()
        .map(|s| s.parse::<ID>().unwrap())
        .collect::<Vec<ID>>();

    let mut maps: Vec<AlmMap> = vec![];
    loop {
        let sect = sections.next();
        if sect.is_none() {
            break;
        }

        let mut sect_iter = sect.unwrap().lines();
        let title = sect_iter.next().unwrap(); // title line: "seed-to-soil map:"
        let config_nums = sect_iter.map(|line| {
            let triplet = line
                .split_whitespace()
                .map(|s| s.parse::<ID>().unwrap())
                .collect::<Vec<ID>>();
            if triplet.len() != 3 {
                panic!("triplet.len() != 3: {} {}", title, line);
            }
            (triplet[0], triplet[1], triplet[2])
        });
        maps.push(AlmMap {
            title: title.to_string(),
            config: config_nums.collect(),
        });
    }
    Almanac { seeds, maps }
}

/// The rest of the almanac contains a list of maps which describe how to convert numbers from a source category into numbers in a destination category.
/// That is, the section that starts with seed-to-soil map: describes how to convert a seed number (the source) to a soil number (the destination).
/// This lets the gardener and his team know which soil to use with which seeds, which water to use with which fertilizer, and so on.
///
/// Rather than list every source number and its corresponding destination number one by one, the maps describe entire ranges of numbers that can be converted.
/// Each line within a map contains three numbers: the destination range start, the source range start, and the range length.
///
/// Consider again the example seed-to-soil map:
/// ```
/// 50 98 2
/// 52 50 48
/// ```
/// The first line has a destination range start of 50, a source range start of 98, and a range length of 2.
/// This line means that the source range starts at 98 and contains two values: 98 and 99.
/// The destination range is the same length, but it starts at 50, so its two values are 50 and 51.
/// With this information, you know that seed number 98 corresponds to soil number 50 and that seed number 99 corresponds to soil number 51.
///
/// The second line means that the source range starts at 50 and contains 48 values: 50, 51, ..., 96, 97.
/// This corresponds to a destination range starting at 52 and also containing 48 values: 52, 53, ..., 98, 99.
/// So, seed number 53 corresponds to soil number 55.
///
/// Any source numbers that aren't mapped correspond to the same destination number.
/// So, seed number 10 corresponds to soil number 10.
pub fn get_dest_by_src(src: &ID, amap: &AlmMap) -> ID {
    // let mut dest = *src;
    let dest = *src;
    for (dest_start, src_start, len) in amap.config.iter() {
        if src_start <= &dest && dest < *src_start + *len {
            // dest = dest_start + val - src_start;
            return dest_start + dest - src_start;
        }
    }
    dest
}

pub fn get_destination(seed: &ID, alm: &Almanac) -> ID {
    let mut val = *seed;
    for map in alm.maps.iter() {
        val = get_dest_by_src(&val, map);
    }
    val
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    const AL_STR: &str = r#"
seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
    "#;

    fn get_almanac() -> Almanac {
        parse_almanac(AL_STR.trim())
    }

    #[test]
    fn test_run() {
        assert_debug_snapshot!(get_almanac());
    }

    fn get_destination_path(seed: &ID, alm: &Almanac) -> Vec<ID> {
        let mut val = vec![*seed];
        for map in alm.maps.iter() {
            val.push(get_dest_by_src(val.last().unwrap(), map));
        }
        val
    }

    #[test]
    fn get_destination_test() {
        // Seed 79, soil 81, fertilizer 81, water 81, light 74, temperature 78, humidity 78, location 82.
        assert_eq!(
            get_destination_path(&79, &get_almanac()),
            [79, 81, 81, 81, 74, 78, 78, 82]
        );
        // Seed 14, soil 14, fertilizer 53, water 49, light 42, temperature 42, humidity 43, location 43.
        assert_eq!(
            get_destination_path(&14, &get_almanac()),
            [14, 14, 53, 49, 42, 42, 43, 43]
        );
        // Seed 55, soil 57, fertilizer 57, water 53, light 46, temperature 82, humidity 82, location 86.
        assert_eq!(
            get_destination_path(&55, &get_almanac()),
            [55, 57, 57, 53, 46, 82, 82, 86]
        );
        // Seed 13, soil 13, fertilizer 52, water 41, light 34, temperature 34, humidity 35, location 35.
        assert_eq!(
            get_destination_path(&13, &get_almanac()),
            [13, 13, 52, 41, 34, 34, 35, 35]
        );
    }
}
