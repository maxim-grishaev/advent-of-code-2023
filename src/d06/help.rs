pub type Time = u32; // keep them separate
pub type Distance = u64; // just in case
pub type Race = (Time, Distance);
pub type Timetable = Vec<Race>;

const TIMETABLE: &str = r#"
Time:        48     98     90     83
Distance:   390   1103   1112   1360
"#;
//4_294_967_295
// const x: (Time, Distance) = (48_989_083, 390_110_311_121_360);

pub fn read_timetable() -> String {
    TIMETABLE.trim().to_string()
}

fn read_line<T>(line_opt: Option<&str>) -> Vec<T>
where
    T: std::str::FromStr,
    <T as std::str::FromStr>::Err: std::fmt::Debug,
{
    line_opt
        .unwrap()
        .split_whitespace()
        .skip(1)
        .map(|s| s.parse::<T>().unwrap())
        .collect()
}

pub fn parse_timetable(timetable: &str) -> Timetable {
    let mut lines = timetable.lines();
    let times = read_line::<Time>(lines.next());
    let distances = read_line::<Distance>(lines.next());
    times
        .into_iter()
        .zip(distances.into_iter())
        .collect::<Timetable>()
}

fn read_line_as_one_number(line_opt: Option<&str>) -> String {
    // println!("{:?}", line_opt);
    line_opt
        .unwrap()
        .split_whitespace()
        .skip(1)
        .collect::<Vec<&str>>()
        .join("")
}

pub fn parse_as_one_race(timetable: &str) -> Race {
    let mut lines = timetable.lines();
    let time = read_line_as_one_number(lines.next())
        .parse::<Time>()
        .unwrap();
    let distance = read_line_as_one_number(lines.next())
        .parse::<Distance>()
        .unwrap();
    (time, distance)
}

fn get_distance_by_charge_time(charge_time: Time, race_time: Time) -> Distance {
    Distance::saturating_mul(
        charge_time as Distance,
        (race_time - charge_time) as Distance,
    )
}

pub fn find_all_race_times(race_time: &Time, record_distance: &Distance) -> Vec<Distance> {
    let mut res: Vec<Distance> = vec![];
    println!(
        "race_time: {:?}, record_distance: {:?}",
        race_time, record_distance
    );
    let mut max_d: Distance = 0;
    for time in 1..*race_time {
        let d = get_distance_by_charge_time(time, *race_time);
        if d > max_d {
            max_d = d;
        }
        if d > *record_distance {
            res.push(d);
        }
    }
    println!("max: {:?}, {:?}", max_d, record_distance);
    println!("tot: {:?}", res.len());
    res
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    const TIMETABLE_TEST: &str = r#"
Time:      7  15   30
Distance:  9  40  200
    "#;

    fn get_ttbl() -> String {
        TIMETABLE_TEST.trim().to_string()
    }

    #[test]
    fn smoke() {
        let timetable = get_ttbl();
        let parsed = parse_timetable(&timetable);
        assert_debug_snapshot!(parsed);
    }

    #[test]
    fn get_times() {
        let timetable = get_ttbl();
        let parsed = parse_timetable(&timetable);
        assert_debug_snapshot!(find_all_race_times(&parsed[0].0, &parsed[0].1));
        assert_debug_snapshot!(find_all_race_times(&parsed[1].0, &parsed[1].1));
        assert_debug_snapshot!(find_all_race_times(&parsed[2].0, &parsed[2].1));
    }

    #[test]
    fn parse_as_one_race_smoke() {
        let timetable = get_ttbl();
        assert_eq!(parse_as_one_race(&timetable), (71530, 940200));
    }

    #[test]
    fn parse_as_one_race_count_times() {
        let ttbl_s = get_ttbl();
        let race = parse_as_one_race(&ttbl_s);
        // println!("{:?}", race); // (71530, 940200)
        assert_eq!(find_all_race_times(&race.0, &race.1).len(), 71_503);
    }
}
