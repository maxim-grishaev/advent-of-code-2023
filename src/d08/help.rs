use crate::util;

use super::map::Map;

pub fn read_map() -> String {
    util::fs::read_file(file!(), "map.txt")
}

pub fn get_primes(n: u128) -> Vec<u128> {
    let mut primes = Vec::new();
    let mut num = n;
    let mut i = 2;

    while i <= num {
        if num % i == 0 {
            primes.push(i);
            num /= i;
        } else {
            i += 1;
        }
    }
    primes
}

pub fn count_steps<IE>(str: &str, start_node: &str, is_end: IE) -> usize
where
    IE: Fn(&str) -> bool,
{
    let map = Map::from(str);
    let dln = map.dir.len();

    let mut i: usize = 0;
    let mut node = start_node;

    loop {
        let d = map.dir.chars().nth(i % dln).unwrap();

        // println!("{:?}", (i, node, d));
        let (l, r) = map.nodes.get(node).unwrap();

        node = match d {
            Map::LEFT => l,
            Map::RIGHT => r,
            _ => panic!("Unknown direction"),
        };

        i += 1;
        if is_end(node) {
            return i;
        }
    }
}

#[cfg(test)]
pub mod t {
    // use super::*;
    use super::super::map::Map;
    use insta::assert_debug_snapshot;

    pub const MAP_1: &str = r#"
RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)
    "#;

    pub const MAP_2: &str = r#"
LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)
    "#;

    //     pub const MAP_GHOST: &str = r#"
    // LR

    // 11A = (11B, XXX)
    // 11B = (XXX, 11Z)
    // 11Z = (11B, XXX)
    // 22A = (22B, XXX)
    // 22B = (22C, 22C)
    // 22C = (22Z, 22Z)
    // 22Z = (22B, 22B)
    // XXX = (XXX, XXX)
    //     "#;

    #[ignore = "snap + hashmap"]
    #[test]
    fn parse() {
        assert_debug_snapshot!(Map::from(MAP_1.trim()));
        assert_debug_snapshot!(Map::from(MAP_2.trim()));
    }

    // #[ignore = "snap + hashmap"]
    // #[test]
    // fn map_1() {
    //     assert_eq!(count_steps(MAP_1.trim()), 2);
    // }

    // #[ignore = "snap + hashmap"]
    // #[test]
    // fn map_2() {
    //     assert_eq!(count_steps(MAP_2.trim()), 6);
    // }
}
