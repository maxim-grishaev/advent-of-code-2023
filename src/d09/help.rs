use std::fmt::Debug;

use crate::util;

pub type Num = i64;
pub type Sum = i128;

#[derive(Clone)]
pub struct Line(Vec<Num>);

impl Debug for Line {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let nums = self
            .0
            .iter()
            .map(|n| n.to_string())
            .collect::<Vec<_>>()
            .join(" ");
        write!(f, "{}", nums)
    }
}

impl Line {
    fn len(&self) -> usize {
        self.0.len()
    }

    fn iter(&self) -> impl Iterator<Item = &Num> {
        self.0.iter()
    }

    fn get(&self, i: usize) -> Option<&Num> {
        self.0.get(i)
    }
}

pub fn read_data() -> String {
    util::fs::read_file(file!(), "readings.txt")
}

pub fn parse_line(str: &str) -> Line {
    Line(str.split_whitespace().map(|n| n.parse().unwrap()).collect())
}

pub fn parse_to_lines(str: &str) -> Vec<Line> {
    str.trim().lines().map(parse_line).collect()
}

pub fn get_diff_line(line: &Line) -> Line {
    let dln = line.len() - 1;
    let mut diffs = vec![0; dln];
    for i in 0..dln {
        diffs[i] = line.get(i + 1).unwrap() - line.get(i).unwrap();
    }
    Line(diffs)
}

pub fn are_zeros(line: &Line) -> bool {
    line.iter().all(|&n| n == 0)
}

pub fn get_tower(line: &Line) -> Vec<Line> {
    let mut diffs: Vec<Line> = vec![];
    let mut cur_line = line.clone();

    while !are_zeros(&cur_line) {
        diffs.push(cur_line.clone());
        cur_line = get_diff_line(&cur_line);
    }

    diffs
}

pub fn count_prev_by_tower(tower: &Vec<Line>) -> Sum {
    let mut count: Sum = 0;
    for line in tower.iter().rev() {
        let first_n = *line.iter().nth(0).unwrap();
        count = first_n as Sum - count;
    }

    count
}

pub fn count_next_by_tower(tower: &Vec<Line>) -> Sum {
    let mut count: Sum = 0;
    for line in tower.iter().rev() {
        let last_n = *line.iter().last().unwrap();
        count += last_n as Sum;
    }

    count
}

#[cfg(test)]
pub mod t {
    use super::*;
    use insta::assert_debug_snapshot;

    pub const EXAMPLE: &str = r#"
0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45
    "#;

    #[test]
    fn test_run() {
        assert_debug_snapshot!(parse_to_lines(EXAMPLE.trim()));
    }

    #[test]
    fn test_run_diffs() {
        let nums = parse_to_lines(EXAMPLE.trim());
        assert_debug_snapshot!(nums.iter().map(get_tower).collect::<Vec<_>>());
    }
}
