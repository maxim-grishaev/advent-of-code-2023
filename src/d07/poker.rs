use std::{collections::HashMap, fmt::Debug};

pub type CardsMap = [char; 13];
pub type Bid = u16;

pub struct Card {
    pub chr: char,
}

impl Debug for Card {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Card {}", &self.chr)
    }
}

impl Card {
    pub fn new(chr: char) -> Self {
        Card { chr }
    }
}

#[derive(Debug)]
pub struct CardsStack {
    pub stack: [Card; 5],
}

impl CardsStack {
    pub fn new(cards_str: &str) -> Self {
        let crd = cards_str.chars().map(Card::new).collect::<Vec<Card>>();
        CardsStack {
            stack: crd.try_into().unwrap(),
        }
    }
    pub fn count_cards(&self) -> Vec<(char, u8)> {
        let mut hm: HashMap<char, u8> = HashMap::new();
        self.stack.iter().for_each(|c| {
            let counter = hm.entry(c.chr).or_insert(0);
            *counter += 1;
        });
        let mut pairs: Vec<(char, u8)> = hm.into_iter().collect();
        pairs.sort_by(|a, b| a.1.cmp(&b.1));
        pairs
    }
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum HandType {
    FiveOfAKind = 7,
    FourOfAKind = 6,
    FullHouse = 5,
    ThreeOfAKind = 4,
    TwoPairs = 3,
    OnePair = 2,
    HighCard = 1,
}

impl HandType {
    pub fn from_sorted_count(counts: Vec<&u8>) -> Self {
        match counts.as_slice() {
            [5] => HandType::FiveOfAKind,
            [1, 4] => HandType::FourOfAKind,
            [2, 3] => HandType::FullHouse,
            [1, 1, 3] => HandType::ThreeOfAKind,
            [1, 2, 2] => HandType::TwoPairs,
            [1, 1, 1, 2] => HandType::OnePair,
            [1, 1, 1, 1, 1] => HandType::HighCard,
            _ => panic!("Unexpected card stack: {:?}", counts),
        }
    }
    pub fn from_sorted_pairs(pairs: Vec<(char, u8)>) -> Self {
        let counts = pairs
            .iter()
            .map(|(_chr, count)| count)
            .collect::<Vec<&u8>>();
        HandType::from_sorted_count(counts)
    }
}

impl From<&CardsStack> for HandType {
    fn from(card_stack: &CardsStack) -> Self {
        HandType::from_sorted_pairs(card_stack.count_cards())
    }
}

#[derive(Debug)]
pub struct GameItem {
    pub card_stack: CardsStack,
    pub bid: Bid,
}

impl GameItem {
    pub fn new(s: &str) -> Self {
        let mut line_split = s.trim().split_whitespace();

        let cards_str = line_split.next().unwrap();
        let card_stack = CardsStack::new(cards_str);
        let bid = line_split.next().unwrap().parse::<Bid>().unwrap();
        GameItem { card_stack, bid }
    }
}

#[cfg(test)]
mod tests {
    use super::super::help;
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn smoke_test() {
        let games = help::t::get_test()
            .lines()
            .map(|l| GameItem::new(l))
            .collect::<Vec<GameItem>>();
        assert_debug_snapshot!(games);
    }
}
