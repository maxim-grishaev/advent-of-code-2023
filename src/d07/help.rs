use super::poker;
use crate::util;

pub fn read_poker_string() -> String {
    util::fs::read_file(file!(), "cards.txt")
}

pub fn parse_poker_string(str: &str) -> Vec<poker::GameItem> {
    str.lines().map(|l| poker::GameItem::new(l)).collect()
}

#[cfg(test)]
pub mod t {
    const STR: &str = r#"
32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483
"#;

    pub fn get_test() -> String {
        STR.trim().to_string()
    }
}
