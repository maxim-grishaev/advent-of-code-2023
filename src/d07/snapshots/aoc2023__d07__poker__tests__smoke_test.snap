---
source: src/d07/poker.rs
expression: games
---
[
    GameItem {
        card_stack: CardsStack {
            stack: [
                Card 3,
                Card 2,
                Card T,
                Card 3,
                Card K,
            ],
        },
        bid: 765,
    },
    GameItem {
        card_stack: CardsStack {
            stack: [
                Card T,
                Card 5,
                Card 5,
                Card J,
                Card 5,
            ],
        },
        bid: 684,
    },
    GameItem {
        card_stack: CardsStack {
            stack: [
                Card K,
                Card K,
                Card 6,
                Card 7,
                Card 7,
            ],
        },
        bid: 28,
    },
    GameItem {
        card_stack: CardsStack {
            stack: [
                Card K,
                Card T,
                Card J,
                Card J,
                Card T,
            ],
        },
        bid: 220,
    },
    GameItem {
        card_stack: CardsStack {
            stack: [
                Card Q,
                Card Q,
                Card Q,
                Card J,
                Card A,
            ],
        },
        bid: 483,
    },
]
